### ISSUE TITLE
Components for the "basic" bootstrap components

### ISSUE THREAD

**Issue by lougreenwood** [kaliber5/ember-bootstrap#694]
I asked this in Discord, but I'm curious to know if there's any appetite to add components to ember-bootstrap for the more "basic" bootstrap components, such as badges, tags, breadcrumbs, cards, tables etc?

I will likely need to build these bootstrap components for my project and if they would be welcome by the project I could contribute them.

**Comment by simonihmig** [comment]
I thought about this in the past, but I am not sure if that's really that useful. Currently there are components mainly for things that need JavaScript (e.g. there is a jquery plugin also for the original Bootstrap).

So what benefit would you see with something like `{{#bs-badge type="secondary"}}New{{/bs-badge}}` compared to plain markup like `<span class="badge badge-secondary">New</span>`? At the end the component would just render the latter, and not provide anything useful on top.

As downsides I see that this increases the size (in KB) of the library, has a performance cost (imagine a whole lot of badges on same table or so) and customizing the elements (like attaching event listeners, custom HTML attributes) could become less flexible.

The size and performance cost could be mitigated by doing the component -> markup conversion not by rendering a runtime component, but instead at build time. https://github.com/cibernox/ember-ast-helpers would support this, although I haven't tried this yet, so not sure how easy and stable that is. @cibernox thoughts?

**Comment by bravo-kernel** [comment]
I second @simonihmig here. As an Ember newbie I initially was looking for components to generate badges and stuff as well but now that I see the simplicity of just creating them I wonder if the overhead of maintenance, testing, size etc. is worth the trouble. FWIW I no longer feel it does.

**Comment by lougreenwood** [comment]
Let me first explain my situation to provide some context...

So I'm currently working on a project to build out an internal UI framework for my employer, the framework is based off of `ember-bootstrap`.

A large part of the work for this is to customise the look & feel - so CSS work. My initial goal was to use BEM & ITCSS patterns to organise the CSS (un-organised CSS is just waiting to be corrupted, corrupted CSS churns and causes many many UI bugs & regressions)... however, it soon became obvious that is not possible, because:

* If we're going to organise around BEM naming, then we should really use BEM-only everywhere (i.e no bootstrap classes), otherwise this is fragmentation which will corrupt over time (and CSS corrupts easily and quickly)
* `Bootstrap` provides nothing like an API for extension (perhaps if all `bootstrap` components were composed from mixing this would be possible, but it's not :) )

So a solution I arrived at was to use `ember-css-modules` to organise my CSS - this provides the benefits that:

* all CSS is organised around the component (this is actually an improvement over Sass and allows me to use Ember's organisation principles for CSS too - so more clarity & better DX)
* no need to worry about BEM or class names too much since much of the class name is auto-generated by `css-modules` to avoid specificity problems & collisions etc.

Over the past week or so I've been going through the process of creating my own components which extend `ember-bootstrap`, so I've essentially extended every single component and yielded component which `ember-bootstrap` provides in my UI framework, the purpose of this is to allow me to integrate `ember-css-modules` init the components and define my CSS which overrides the base bootstrap styling.

Now... onto basic components...

So, to allow me to continue using `ember-css-modules` for less obvious bootstrap components (badge, tag table etc), I need to create some components for these (the alternative is to use plain CSS to override use bootstrap, but this drops all of benefits of `ember-css-modules` and takes me back to the start).

These components need to be created by me anyway, and for for the sake of consistency, I'll need to create them so they align with `ember-bootstrap` as closely as possible so the usage of my internal UI framework feels homogenous (aka good DX).

So that made me think that I could possibly contribute these base components back to `ember-bootstrap` for other users who may wish to leverage `ember-css-modules` (or any other css framework which relies on class name generation by the component, such as `css-blocks` & maybe even `ember-component-css`).

TLDR;

I need to create components for these "basic" `bootstrap` components so I can extend them in an internal UI framework and use `ember-css-modules`. If I'm creating them internally for a valid use case, perhaps I could contribute some of the components to the project.

:)

**Comment by cibernox** [comment]
I'd say we shouldn't abuse built time transforms. Specially given the goal of the reporter. Personally if those components have no behavior, just html, I usually incline for use plain html

**Comment by simonihmig** [comment]
@lougreenwood thanks for the thorough explanation, makes much more sense now.

I wouldn't be against adding a bunch of components, though again I see not much benefit in them, except for your rather special case. But it probably wouldn't do any harm, especially that we already have some white/blacklist functionality, so users could omit them when they prefer to write plain HTML. It's just that I won't be able to spend much time on this "feature", except for reviewing things and giving a helping hand if required...

But I think we should maybe define the scope of this endeavor. Like there might be blurry lines, and we should be clear about what we want to add and what not.

One possible way to define that scope, and one that I could agree with, could be to simply say anything that Bootstrap itself calls a "component" (see https://getbootstrap.com/docs/4.1/components/alerts/) can be an `ember-bootstrap` component as well, but not more! Does that make sense?

For example I would be worried about trying to add tables as Ember components (i.e. with yielded components for `<tr>`, `<td>` and what not), as tables can come in so many different shapes, that it would be quite easy to have the component approach be very limiting. Just think of all the HTML attributes that people might want to set. Unless they are using angle bracket invocation, the components would need to explicitly enables all of them with `attributeBindings`. And you would often create a huge number of components, certainly with negative performance implications. So I would rather not go into that rabbit hole...

Another thing is the approach to supporting Bootstrap 3. Which I guess you are not planning to do. Which is maybe fine, but as currently all components come with dual BS3/BS4 versions, that would add some inconsistency to the addon, and require at least some proper documentation.
